applylib/applyset/applyset.go

```diff
@@ -153,11 +242,151 @@ func (a *ApplySet) ApplyOnce(ctx context.Context) (*ApplyResults, error) {
+	return nil
+}
+
+// updateParentLabelsAndAnnotations updates the parent labels and annotations.
+// This method leverages the kubectlapply to build the parent labels and annotations, but avoid using the
+// `resource.NewHelper` and cmdutil to patch the parent.
+func (a *ApplySet) updateParentLabelsAndAnnotations(ctx context.Context, kapplyset *kubectlapply.ApplySet, mode kubectlapply.ApplySetUpdateMode) error {
+	parent, err := meta.Accessor(a.parent.GetSubject())
+	if err != nil {
+		return err
+	}
+
+	original, err := meta.Accessor(a.parent.GetSubject().DeepCopyObject())
+	if err != nil {
+		return err
+	}
+	partialParent := kapplyset.BuildParentPatch(mode)
+	newAnnotations := mergeMap(partialParent.Annotations, parent.GetAnnotations())
+	parent.SetAnnotations(newAnnotations)
+	newLabels := mergeMap(partialParent.Labels, parent.GetLabels())
+	parent.SetLabels(newLabels)
+
+	return a.updateParent(ctx, original, parent)
+}
+
+func (a *ApplySet) deleteObjects(ctx context.Context, pruneObjects []kubectlapply.PruneObject, results *ApplyResults) error {
+	for i := range pruneObjects {
+		pruneObject := &pruneObjects[i]
+
+		name := pruneObject.Name
+		namespace := pruneObject.Namespace
+		mapping := pruneObject.Mapping
+		gvk := pruneObject.Object.GetObjectKind().GroupVersionKind()
+		nn := types.NamespacedName{Namespace: namespace, Name: name}
+
+		if err := a.client.Resource(mapping.Resource).Namespace(namespace).Delete(ctx, name, a.deleteOptions); err != nil {
+			klog.Error("unable to delete resource ")
+			results.pruneError(gvk, nn, fmt.Errorf("error from delete: %w", err))
+		} else {
+			klog.Infof("pruned resource %v", pruneObject.String())
+			results.pruneSuccess(gvk, nn)
+		}
+	}
+	return nil
+}
+
+func (a *ApplySet) WithParent(ctx context.Context, kapplyset *kubectlapply.ApplySet) error {
+	parent := a.parent.GetSubject()
+	object, err := meta.Accessor(parent)
+	if err != nil {
+		return err
+	}
+	original, err := meta.Accessor(parent.DeepCopyObject())
+	if err != nil {
+		return err
+	}
+	//kubectlapply requires the tooling and id to exist.
+	{
+		annotations := object.GetAnnotations()
+		if annotations == nil {
+			annotations = make(map[string]string)
+		}
+		annotations[kubectlapply.ApplySetToolingAnnotation] = a.tooling
+		if _, ok := annotations[kubectlapply.ApplySetGRsAnnotation]; !ok {
+			annotations[kubectlapply.ApplySetGRsAnnotation] = ""
+		}
+		object.SetAnnotations(annotations)
+
+		labels := object.GetLabels()
+		if labels == nil {
+			labels = make(map[string]string)
+		}
+		labels[kubectlapply.ApplySetParentIDLabel] = kapplyset.ID()
+		object.SetLabels(labels)
+	}
+	// This is not a cluster fetch. It builds up the parents labels and annotations information in kapplyset.
+	if err := kapplyset.FetchParent(a.parent.GetSubject()); err != nil {
+		return err
+	}
+
+	return a.updateParent(ctx, original, object)
+}
+
+func (a *ApplySet) updateParent(ctx context.Context, original, current metav1.Object) error {
```

Comment:
Maybe rename to updateParentLabelsAndAnnotations ?

And actually if you wanted to, you could inline the functionality of mergeMap into this function, so it would take `labels` and `annotations` as parameters

Reply:
done. merged to updateParentLabelsAndAnnotations because this method now is only called once.